core/backend/src/BackendHubAccess.ts

```diff
@@ -227,6 +227,12 @@ export interface BackendHubAccess {
    */
   queryV2Checkpoint: (arg: CheckpointProps) => Promise<V2CheckpointAccessProps | undefined>;
 
+  /**
+   * Get the access props for a V2 checkpoint that is at the requested changesetId or the most recent checkpoint before it. Returns undefined if no such V2 checkpoint exists.
```

Comment:
This function (not yet implemented) should probably only return v2 checkpoints which are marked as successful. As a result, I should also update this comment to reflect that, does that seem reasonable?

Reply:
How can they get the most recent checkpoint if we've only provided a function to query a specific one? The way its written now, we'll just fallback to a v1 checkpoint if the specific checkpoint they requested is not created and successful. 

We might be able to determine the list of existing checkpoints by reading the manifest, but we'd still need to query to get access to the container from the hub. I also worry about any checkpoints created pre cloudsqlite, which could potentially include failed checkpoints in the manifest. I'd prefer to go by what the hub knows is successful(reported as such by the v2-checkpoint-agent).

Reply:
Let's make a query that just gets the checkpoint container name and access token from the hub. Then, all the logic to determine which checkpoint to download should be on the client.

Reply:
So I guess the clients would be calling queryDatabases on the container? How would a client know which database is the most recent? I'm guessing thats probably the most common case, a client wants the most recent checkpoint. If an app is looking for a specific changeset ID / index, and they're given a list of databases from their queryDatabases and their specific ID isn't in the list, what do they do? 

I worry that we're shifting too much logic onto the clients, as opposed to just having them call CheckpointManager.downloadCheckpoint and ending up with what they requested.

Reply:
i'd far rather put the work on the client (in BriefcaseManager.downloadBriefcase) than on the server. How do we name checkpoints, by changeset id or index?

Reply:
Ah sorry I was equating "client" with consumers of the itwinjs core packages. A CheckpointV2 instance returned by the hub only has the changesetId on it. Thats using the internal APIs though, not sure if its any different for the public imodelhub ones. I'd have to ask someone on their team.  From the perspective of the manifest, entries also have dbName as changesetId. Not sure if that answers your question or not..

Reply:
that's unfortunate, I thought I suggested we use changeset index for the dbName? Really, changesetId is a useless concept that should be eliminated (indexes can be sorted, ids can't). 

I guess the thing to do is to change `CheckpointProps.changeset` member to allow undefined, and then have `queryV2Checkpoint`  return the latest checkpoint if it's not supplied.

Reply:
That would probably work for most use cases, but I still like the flexibility of adding queryCurrentOrPreceding. Can we say for sure that the latest checkpoint will satisfy everyone's needs? I was hoping we could add this as optional here and then make it a breaking change for 4.0.

Reply:
cc @robertas-kerpys @austeja-bentley

Reply:
>  Can we say for sure that the latest checkpoint will satisfy everyone's needs?

Yes, I can say for certain that I *don't* want anyone creating a briefcase other than from the latest checkpoint.

Reply:
@kabentley What about if I just want to download and create a snapshot?  Since you can't "unapply" a schema changes, I have to figure we'll always need _some_ way to find a changeset's "nearest" prior checkpoint...

Also, I did want to clarify, this is not something we're proposing _adding_ to the REST API - this exists _today_.  When I [GET a changeset](https://developer.bentley.com/apis/imodels-v2/operations/get-imodel-changeset-details/) from the iModels API, part of the response payload is the ID of the `currentOrPrecedingCheckpoint`.  So the change here is more about letting us surface that existing info via this client interface...

Reply:
The change I suggest (to allow `undefined` for `CheckpointProps.changeset`) is the right thing to do, since we need always need to query to get the `CheckpointAccessProps` to create a briefcase. That covers that 99.99% use case.

For the very rare case of creating a snapshot from a point on the timeline other than an existing checkpoint, there does need to be a way to correlate checkpoints with the timeline. This will continue to be a problem as long as ChangeSetId is used in our apis. Why did we use changesetId as the identifier for checkpoints? Can we change that?

Mainly, navigating through the available checkpoints should be authoritatively determined by the manifest in the checkpoint's container. That is the *real* answer to what's available, and it is free and fast for any client to read. You should be able to sort the checkpoints from the manifest itself.